operations/iuf/workflows/upgrade_all_products.md

```diff
@@ -392,20 +392,250 @@ Once this step has completed:
-remember to include these workflows.
+**IMPORTANT** There is a different procedure for  `management-nodes-rollout` depending on whether or not CSM is being upgraded.
+The two procedures differ in the handling of NCN storage nodes and NCN master nodes. If CSM is not being upgraded, then NCN storage nodes and NCN master nodes will not be upgraded and will be updated by the CFS configuration
+created in [update-cfs-config](../stages/update_cfs_config.md).
+If CSM is being upgraded, the NCN storage nodes and NCN master nodes will be upgraded.
+Both procedures use the same steps for rebuilding/upgrading NCN worker nodes.
+Select **one** of the following procedures based on whether or not CSM is being upgraded.
+
+- [Management-nodes-rollout with CSM upgrade](#651-management-nodes-rollout-with-csm-upgrade)
+- [Management-nodes-rollout with no CSM upgrade](#652-management-nodes-rollout-with-no-csm-upgrade)
+
+#### 6.5.1 Management-nodes-rollout with CSM upgrade
+
+All management nodes will need to be upgraded to a new image because CSM is being upgraded. NCN master nodes, excluding `ncn-m001`, and NCN worker nodes can be upgraded with IUF.
+NCN storage nodes and `ncn-m001` will be upgraded with manual commands.
+This section describes how to test a new image and CFS configuration on a single "canary node" for NCN master nodes and NCN worker nodes first before rolling it out to the other NCN master nodes and NCN worker nodes.
+Follow the steps below to upgrade all management nodes.
+
+1. Refer to the "Install and Upgrade Framework" section of each individual product's installation documentation to determine if any special actions need to be performed outside of IUF for the `management-nodes-rollout` stage.
+
+1. Get the image-id and CFS configuration created during `prepare-images` and `update-cfs-config` stages.
+Follow the instructions in [prepare-images](../stages/prepare_images.md#artifacts-created) to get the artifacts for `management-node-images`. For the images with the `configuration_group_name` matching
+`Management_Master` and `Management_Storage`, get the values for `final_image_id` and `configuration`.
+These values will be needed when upgrading NCN storage nodes and `ncn-m001`.
+
+1. NCN storage node upgrade
+
+    1. Set the CFS configuration on all storage nodes.
+
+        1. (`ncn-m#`) Set `CFS_CONFIG_NAME` to be the value for `configuration` found for `Management_Storage` nodes in the previous step.
+
+            ```bash
+            CFS_CONFIG_NAME=configuration
+            ```
+
+        1. (`ncn-m#`) Get all NCN storage node xnames.
+
+            ```bash
+            XNAMES=$(cray hsm state components list --role Management --subrole Storage --type Node --format json | jq -r '.Components | map(.ID) | join(",")')
+            echo $XNAMES
```

Comment:
Mitch told me to document all variables like ```"${XNAMES}"```.  This change applies here and in other places below where $XNAMES and $XNAME are referred to.